tiledb/api/src/array/enumeration.rs

```diff
@@ -0,0 +1,502 @@
+    pub(crate) context: &'ctx Context,
+    pub(crate) raw: RawEnumeration,
+}
+
+impl<'ctx> Enumeration<'ctx> {
+    pub(crate) fn capi(&self) -> *mut ffi::tiledb_enumeration_t {
+        *self.raw
+    }
+
+    pub fn name(&self) -> TileDBResult<String> {
+        let mut c_str: *mut ffi::tiledb_string_t = out_ptr!();
+        let res = unsafe {
+            ffi::tiledb_enumeration_get_name(
+                self.context.capi(),
+                self.capi(),
+                &mut c_str,
+            )
+        };
+        if res == ffi::TILEDB_OK {
+            let tdb_str = TDBString {
+                raw: RawTDBString::Owned(c_str),
+            };
+            tdb_str.to_string()
+        } else {
+            Err(self.context.expect_last_error())
+        }
+    }
+
+    pub fn datatype(&self) -> TileDBResult<Datatype> {
+        let mut dtype: ffi::tiledb_datatype_t = 0;
+        let res = unsafe {
+            ffi::tiledb_enumeration_get_type(
+                self.context.capi(),
+                self.capi(),
+                &mut dtype,
+            )
+        };
+        if res == ffi::TILEDB_OK {
+            Ok(Datatype::from_capi_enum(dtype))
+        } else {
+            Err(self.context.expect_last_error())
+        }
+    }
+
+    pub fn cell_val_num(&self) -> TileDBResult<u32> {
+        let mut c_cvn: u32 = 0;
+        let res = unsafe {
+            ffi::tiledb_enumeration_get_cell_val_num(
+                self.context.capi(),
+                self.capi(),
+                &mut c_cvn,
+            )
+        };
+        if res == ffi::TILEDB_OK {
+            Ok(c_cvn)
+        } else {
+            Err(self.context.expect_last_error())
+        }
+    }
+
+    pub fn is_var_sized(&self) -> TileDBResult<bool> {
+        Ok(self.cell_val_num()? == u32::MAX)
+    }
+
+    pub fn ordered(&self) -> TileDBResult<bool> {
+        let mut c_ordered: i32 = 0;
+        let res = unsafe {
+            ffi::tiledb_enumeration_get_ordered(
+                self.context.capi(),
+                self.capi(),
+                &mut c_ordered,
+            )
+        };
+        if res == ffi::TILEDB_OK {
+            Ok(c_ordered != 0)
+        } else {
+            Err(self.context.expect_last_error())
+        }
+    }
+
+    pub fn data(&self) -> TileDBResult<&[u8]> {
+        let mut ptr: *const std::ffi::c_uchar = out_ptr!();
+        let mut size: u64 = 0;
+        let res = unsafe {
+            ffi::tiledb_enumeration_get_data(
+                self.context.capi(),
+                self.capi(),
+                &mut ptr as *mut *const std::ffi::c_uchar
+                    as *mut *const std::ffi::c_void,
+                &mut size,
+            )
+        };
+
+        // Rust docs say that we're not allowed to pass a nullptr to the
+        // std::slice::from_raw_parts because that breaks how Option<&[T]>::None
+        // is represented.
+        if ptr.is_null() {
+            ptr = 0x1 as *const std::ffi::c_uchar;
```

Comment:
Doesn't this all but guarantee a segfault?  I guess not because the size is zero?

This line gave me a lot of pause, but it does turn out there is an idiomatic way to do this mentioned on the [function reference page](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html):

"Data must be non-null and aligned even for zero-length slices. One reason for this is that enum layout optimizations may rely on references (including slices of any length) being aligned and non-null to distinguish them from other data. You can obtain a pointer that is usable as data for zero-length slices using [NonNull::dangling()](https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.dangling)."

Reply:
Oooh, good find on that `NonNull::dangling()` to clean this up. I had seen the note about 0x0 vs 0x1 issue so just implemented it directly figuring that all the alignment stuff would Just Work since it was Rust controlling things. However, using the named stdlib approach is even better than the weird looking pointer cast that as you pointed out looks exactly like prime segfault territory even though Rust actually requires it.

I assume you're on board with just doing:

```rust
if ptr.is_null() {
    ptr = std::ptr::NonNull::dangling();
}
```

Reply:
Yep that works for me